File: src/main/java/com/manson/fo76/domain/dto/ItemDTO.java
package com.manson.fo76.domain.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonProperty.Access;
import com.manson.fo76.domain.items.enums.FilterFlag;
import java.util.List;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;

@JsonIgnoreProperties(ignoreUnknown = true)
public class ItemDTO {

  @Id
  private String id;
  @Indexed
  private String ownerId;
  @Indexed
  private String ownerName;
  @Indexed
  private String description;
  private Double price;
  @Indexed
  private Boolean tradeOnly;
  @Indexed
  private Boolean tradePossible;
  @Indexed
  private String text;
  @Indexed(unique = true)
  @JsonProperty(access = Access.WRITE_ONLY)
  private Long serverHandleId;
  private Integer count;
  private Integer itemValue;
  @Indexed
  private FilterFlag filterFlag;
  private Integer currentHealth;
  private Integer damage;
  private Integer durability;
  private Integer maximumHealth;
  private Double weight;
  private Double weaponDisplayAccuracy;
  private Double weaponDisplayRateOfFire;
  private Double weaponDisplayRange;
  @Indexed
  private Integer numLegendaryStars;
  @Indexed
  private Integer itemLevel;
  private Integer rarity;
  private Boolean isTradable;
  private Boolean isSpoiled;
  private Boolean isSetItem;
  private Boolean isQuestItem;
  @Indexed
  private Boolean isLegendary;
  @Indexed
  private List<StatsDTO> stats;
  @Indexed
  private List<LegendaryMod> legendaryMods;
  // TODO: Remove this, once final UI will be ready
  private List<String> legendaryModsTemp;

  public List<String> getLegendaryModsTemp() {
    return legendaryModsTemp;
  }

  public void setLegendaryModsTemp(List<String> legendaryModsTemp) {
    this.legendaryModsTemp = legendaryModsTemp;
  }

  public List<LegendaryMod> getLegendaryMods() {
    return legendaryMods;
  }

  public void setLegendaryMods(List<LegendaryMod> legendaryMods) {
    this.legendaryMods = legendaryMods;
  }

  public List<StatsDTO> getStats() {
    return stats;
  }

  public void setStats(List<StatsDTO> stats) {
    this.stats = stats;
  }

  public String getOwnerName() {
    return ownerName;
  }

  public void setOwnerName(String ownerName) {
    this.ownerName = ownerName;
  }

  public String getId() {
    return id;
  }

  public void setId(String id) {
    this.id = id;
  }

  public String getOwnerId() {
    return ownerId;
  }

  public void setOwnerId(String ownerId) {
    this.ownerId = ownerId;
  }

  public String getDescription() {
    return description;
  }

  public void setDescription(String description) {
    this.description = description;
  }

  public Double getPrice() {
    return price;
  }

  public void setPrice(Double price) {
    this.price = price;
  }

  public Boolean getTradeOnly() {
    return tradeOnly;
  }

  public void setTradeOnly(Boolean tradeOnly) {
    this.tradeOnly = tradeOnly;
  }

  public Boolean getTradePossible() {
    return tradePossible;
  }

  public void setTradePossible(Boolean tradePossible) {
    this.tradePossible = tradePossible;
  }

  public String getText() {
    return text;
  }

  public void setText(String text) {
    this.text = text;
  }

  public Long getServerHandleId() {
    return serverHandleId;
  }

  public void setServerHandleId(Long serverHandleId) {
    this.serverHandleId = serverHandleId;
  }

  public Integer getCount() {
    return count;
  }

  public void setCount(Integer count) {
    this.count = count;
  }

  public Integer getItemValue() {
    return itemValue;
  }

  public void setItemValue(Integer itemValue) {
    this.itemValue = itemValue;
  }

  public FilterFlag getFilterFlag() {
    return filterFlag;
  }

  public void setFilterFlag(FilterFlag filterFlag) {
    this.filterFlag = filterFlag;
  }

  public Integer getCurrentHealth() {
    return currentHealth;
  }

  public void setCurrentHealth(Integer currentHealth) {
    this.currentHealth = currentHealth;
  }

  public Integer getDamage() {
    return damage;
  }

  public void setDamage(Integer damage) {
    this.damage = damage;
  }

  public Integer getDurability() {
    return durability;
  }

  public void setDurability(Integer durability) {
    this.durability = durability;
  }

  public Integer getMaximumHealth() {
    return maximumHealth;
  }

  public void setMaximumHealth(Integer maximumHealth) {
    this.maximumHealth = maximumHealth;
  }

  public Double getWeight() {
    return weight;
  }

  public void setWeight(Double weight) {
    this.weight = weight;
  }

  public Double getWeaponDisplayAccuracy() {
    return weaponDisplayAccuracy;
  }

  public void setWeaponDisplayAccuracy(Double weaponDisplayAccuracy) {
    this.weaponDisplayAccuracy = weaponDisplayAccuracy;
  }

  public Double getWeaponDisplayRateOfFire() {
    return weaponDisplayRateOfFire;
  }

  public void setWeaponDisplayRateOfFire(Double weaponDisplayRateOfFire) {
    this.weaponDisplayRateOfFire = weaponDisplayRateOfFire;
  }

  public Double getWeaponDisplayRange() {
    return weaponDisplayRange;
  }

  public void setWeaponDisplayRange(Double weaponDisplayRange) {
    this.weaponDisplayRange = weaponDisplayRange;
  }

  public Integer getNumLegendaryStars() {
    return numLegendaryStars;
  }

  public void setNumLegendaryStars(Integer numLegendaryStars) {
    this.numLegendaryStars = numLegendaryStars;
  }

  public Integer getItemLevel() {
    return itemLevel;
  }

  public void setItemLevel(Integer itemLevel) {
    this.itemLevel = itemLevel;
  }

  public Integer getRarity() {
    return rarity;
  }

  public void setRarity(Integer rarity) {
    this.rarity = rarity;
  }

  public Boolean getTradable() {
    return isTradable;
  }

  public void setTradable(Boolean tradable) {
    isTradable = tradable;
  }

  public Boolean getSpoiled() {
    return isSpoiled;
  }

  public void setSpoiled(Boolean spoiled) {
    isSpoiled = spoiled;
  }

  public Boolean getSetItem() {
    return isSetItem;
  }

  public void setSetItem(Boolean setItem) {
    isSetItem = setItem;
  }

  public Boolean getQuestItem() {
    return isQuestItem;
  }

  public void setQuestItem(Boolean questItem) {
    isQuestItem = questItem;
  }

  public Boolean getLegendary() {
    return isLegendary;
  }

  public void setLegendary(Boolean legendary) {
    isLegendary = legendary;
  }
}

File: src/main/java/com/manson/fo76/domain/dto/ItemDTO.kt
package com.manson.fo76.domain.dto

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.manson.fo76.domain.items.enums.ArmorType
import com.manson.fo76.domain.items.enums.FilterFlag
import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed

@JsonIgnoreProperties(ignoreUnknown = true)
class ItemDTO {
    @Id
    var id: String? = null

    @Indexed
    var ownerInfo: OwnerInfo? = null

    @Indexed
    var tradeOptions: TradeOptions? = null

    @Indexed
    var text: String? = null

    @Indexed
    var description: String? = null

    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    var serverHandleId: Long = 0
    var count: Int = 0
    var itemValue: Int = 0

    @Indexed
    var filterFlag: FilterFlag = FilterFlag.UNKNOWN
    var currentHealth: Int = 0
    var damage: Int = 0
    var durability: Int = 0
    var maximumHealth: Int = 0
    var weight: Double = 0.0
    var weaponDisplayAccuracy: Double = 0.0
    var weaponDisplayRateOfFire: Double = 0.0
    var weaponDisplayRange: Double = 0.0

    @Indexed
    var numLegendaryStars: Int = 0

    @Indexed
    var itemLevel: Int = 0
    var rarity: Int = 0
    var isTradable: Boolean = false
    var isSpoiled: Boolean = false
    var isSetItem: Boolean = false
    var isQuestItem: Boolean = false

    @Indexed
    var isLegendary: Boolean = false

    @Indexed
    var stats: List<StatsDTO> = listOf()

    @Indexed
    var legendaryMods: List<LegendaryMod> = listOf()
    var abbreviation: String = ""

    var armorType: ArmorType = ArmorType.Unknown
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ItemDTO) return false

        if (id != other.id) return false
        if (ownerInfo != other.ownerInfo) return false
        if (tradeOptions != other.tradeOptions) return false
        if (text != other.text) return false
        if (description != other.description) return false
        if (serverHandleId != other.serverHandleId) return false
        if (count != other.count) return false
        if (itemValue != other.itemValue) return false
        if (filterFlag != other.filterFlag) return false
        if (currentHealth != other.currentHealth) return false
        if (damage != other.damage) return false
        if (durability != other.durability) return false
        if (maximumHealth != other.maximumHealth) return false
        if (weight != other.weight) return false
        if (weaponDisplayAccuracy != other.weaponDisplayAccuracy) return false
        if (weaponDisplayRateOfFire != other.weaponDisplayRateOfFire) return false
        if (weaponDisplayRange != other.weaponDisplayRange) return false
        if (numLegendaryStars != other.numLegendaryStars) return false
        if (itemLevel != other.itemLevel) return false
        if (rarity != other.rarity) return false
        if (isTradable != other.isTradable) return false
        if (isSpoiled != other.isSpoiled) return false
        if (isSetItem != other.isSetItem) return false
        if (isQuestItem != other.isQuestItem) return false
        if (isLegendary != other.isLegendary) return false
        if (stats != other.stats) return false
        if (legendaryMods != other.legendaryMods) return false
        if (abbreviation != other.abbreviation) return false
        if (armorType != other.armorType) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + (ownerInfo?.hashCode() ?: 0)
        result = 31 * result + (tradeOptions?.hashCode() ?: 0)
        result = 31 * result + (text?.hashCode() ?: 0)
        result = 31 * result + (description?.hashCode() ?: 0)
        result = 31 * result + serverHandleId.hashCode()
        result = 31 * result + count
        result = 31 * result + itemValue
        result = 31 * result + filterFlag.hashCode()
        result = 31 * result + currentHealth
        result = 31 * result + damage
        result = 31 * result + durability
        result = 31 * result + maximumHealth
        result = 31 * result + weight.hashCode()
        result = 31 * result + weaponDisplayAccuracy.hashCode()
        result = 31 * result + weaponDisplayRateOfFire.hashCode()
        result = 31 * result + weaponDisplayRange.hashCode()
        result = 31 * result + numLegendaryStars
        result = 31 * result + itemLevel
        result = 31 * result + rarity
        result = 31 * result + isTradable.hashCode()
        result = 31 * result + isSpoiled.hashCode()
        result = 31 * result + isSetItem.hashCode()
        result = 31 * result + isQuestItem.hashCode()
        result = 31 * result + isLegendary.hashCode()
        result = 31 * result + stats.hashCode()
        result = 31 * result + legendaryMods.hashCode()
        result = 31 * result + abbreviation.hashCode()
        result = 31 * result + armorType.hashCode()
        return result
    }


}